packages/cli/src/processed-data-managers/processed-data-helper.ts

```diff
@@ -0,0 +1,322 @@
+			);
+		}
+	}
+
+	private static createContext(
+		context: ProcessedDataContext,
+		contextData: ICheckProcessedContextData,
+	): string {
+		if (context === 'node') {
+			console.log('contextData.node', contextData, contextData.node);
+			if (!contextData.node) {
+				throw new ApplicationError(
+					"No node information has been provided and can so not use context 'node'",
+				);
+			}
+			// Use the node ID to make sure that the data can still be accessed and does not get deleted
+			// whenver the node gets renamed
+			return `n:${contextData.node.id}`;
+		}
+
+		return '';
+	}
+
+	private static createValueHash(value: ProcessedDataItemTypes): string {
+		return createHash('md5').update(value.toString()).digest('base64');
+	}
+
+	async checkProcessed(
+		items: ProcessedDataItemTypes[],
+		context: ProcessedDataContext,
+		contextData: ICheckProcessedContextData,
+		options: ICheckProcessedOptions,
+	): Promise<ICheckProcessedOutput> {
+		const returnData: ICheckProcessedOutput = {
+			new: [],
+			processed: [],
+		};
+
+		const processedData = await Container.get(ProcessedDataRepository).findOne({
+			where: {
+				workflowId: contextData.workflow.id as string,
+				context: ProcessedDataHelper.createContext(context, contextData),
+			},
+		});
+
+		if (processedData && processedData.value.mode !== options.mode) {
+			throw new ApplicationError(
+				'Deduplication data was originally saved with an incompatible setting of the ‘Keep Items Where’ parameter. ’Try Clean Database’ operation to reset.',
+			);
+		}
+
+		if (!processedData) {
+			// If there is nothing it the database all items are new
+			returnData.new = items;
+			return returnData;
+		}
+
+		if (['latestIncrementalKey', 'latestDate'].includes(options.mode)) {
+			const processedDataValue = processedData.value as IProcessedDataLatest;
+
+			const incomingItems = ProcessedDataHelper.sortEntries(items, options.mode);
+			incomingItems.forEach((item) => {
+				if (ProcessedDataHelper.compareValues(options.mode, item, processedDataValue.data)) {
+					returnData.new.push(item);
+				} else {
+					returnData.processed.push(item);
+				}
+			});
+			return returnData;
+		}
+
+		const hashedItems = items.map((item) => ProcessedDataHelper.createValueHash(item));
+
+		hashedItems.forEach((item, index) => {
+			if ((processedData.value.data as string[]).find((entry) => entry === item)) {
+				returnData.processed.push(items[index]);
+			} else {
+				returnData.new.push(items[index]);
+			}
+		});
+
+		return returnData;
+	}
+
+	async checkProcessedAndRecord(
+		items: ProcessedDataItemTypes[],
+		context: ProcessedDataContext,
+		contextData: ICheckProcessedContextData,
+		options: ICheckProcessedOptions,
+	): Promise<ICheckProcessedOutput> {
+		const dbContext = ProcessedDataHelper.createContext(context, contextData);
+
+		if (contextData.workflow.id === undefined) {
+			throw new ApplicationError('Workflow has to have an ID set!');
+		}
```

Comment:
Is this a state that user can create? Or is it a programmer error? If it's the latter, we should use `assert` instead

```suggestion
		a.ok(contextData.workflow.id)
```
with
```ts
import * as a from "node:assert/strict"
```